types.go

```diff
@@ -0,0 +1,296 @@
+package autopilot
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/raft"
+)
+
+//go:generate mockery -all -inpkg -case snake -testonly
+
+// RaftState is the status of a single server in the Raft cluster.
+type RaftState string
+
+const (
+	RaftNone     RaftState = "none"
+	RaftLeader   RaftState = "leader"
+	RaftVoter    RaftState = "voter"
+	RaftNonVoter RaftState = "non-voter"
+	RaftStaging  RaftState = "staging"
+)
+
+func (s RaftState) IsPotentialVoter() bool {
+	switch s {
+	case RaftVoter, RaftStaging, RaftLeader:
+		return true
+	default:
+		return false
+	}
+}
+
+// NodeStatus represents the health of a server as know to the autopilot consumer.
+// This should not take into account Raft health and the server being on a new enough
+// term and index.
+type NodeStatus string
+
+const (
+	NodeUnknown NodeStatus = "unknown"
+	NodeAlive   NodeStatus = "alive"
+	NodeFailed  NodeStatus = "failed"
+	NodeLeft    NodeStatus = "left"
+)
+
+type NodeType string
+
+const (
+	NodeVoter NodeType = "voter"
+)
+
+// Config represents all the tunables of autopilot
+type Config struct {
+	// CleanupDeadServers controls whether to remove dead servers when a new
+	// server is added to the Raft peers.
+	CleanupDeadServers bool
+
+	// LastContactThreshold is the limit on the amount of time a server can go
+	// without leader contact before being considered unhealthy.
+	LastContactThreshold time.Duration
+
+	// MaxTrailingLogs is the amount of entries in the Raft Log that a server can
+	// be behind before being considered unhealthy.
+	MaxTrailingLogs uint64
+
+	// MinQuorum sets the minimum number of servers required in a cluster
+	// before autopilot can prune dead servers.
+	MinQuorum uint
+
+	// ServerStabilizationTime is the minimum amount of time a server must be
+	// in a stable, healthy state before it can be added to the cluster. Only
+	// applicable with Raft protocol version 3 or higher.
+	ServerStabilizationTime time.Duration
```

Comment:
If it wasn't immediately clear the reason for having the `State.ServerStabilizationTime` is to allow for the State to override the user provided value in the case where autopilot hasn't been running long enough for any servers to be able to meet that stabilization time. This prevents unnecessary leadership flapping when autopilot first starts. Only autopilot knows about the time autopilot was started and this value gets put into the autopilot state. The main autopilot type itself also knows about it but the Promoters do not ever get given a handle to the autopilot instance and I wanted to prevent that if possible for circular dependency reasons among other things.

While we could attempt to have promoter specific config, it wouldn't work for the same consistency reasons I mentioned in other comment responses. We must atomically load the whole configuration and use it for all operations during a round of reconciliation or state updating. When means the promoters cannot retrieve it themselves.